packages/js/src/settings/components/search.js

```diff
@@ -29,96 +16,137 @@ const Search = () => {
-		const q = trim( event.target.value );
-
-		if ( q.length > QUERY_MIN_CHARS ) {
-			setResults( groupItemsByRoute(
-				filter( items, item => item?.fieldLabel?.includes( q ) )
-			) );
+		setQuery( "" );
+		setResults( [] );
+	}, [ setClose, setQuery ] );
+
+	const debouncedSearch = useCallback( debounce( newQuery => {
+		const trimmedQuery = trim( newQuery );
+
+		// Bail if query is too short.
+		if ( trimmedQuery.length < QUERY_MIN_CHARS ) {
+			return false;
 		}
 
+		// Split query into words.
+		const splitQuery = split( trimmedQuery, " " );
+
+		// Filter search index by split query and store number of hits.
+		// A hit is registered if a single word from split query in found in a fields keywords.
+		const queryResults = reduce( queryableSearchIndex, ( queryResultsAcc, item ) => {
+			const hits = reduce( splitQuery, ( hitsAcc, queryWord ) => includes( item?.keywords, queryWord ) ? ++hitsAcc : hitsAcc, 0 );
+
+			// Bail if no hits found.
+			if ( hits === 0 ) {
+				return queryResultsAcc;
+			}
+
+			return [
+				...queryResultsAcc,
+				{
+					...item,
+					// Store hits for later sorting.
+					hits,
+				},
+			];
+		}, [] );
+
+		// Sort query results by number of hits on field, highest number of hits first.
+		const sortedQueryResults = queryResults.sort( ( a, b ) => b.hits - a.hits );
+
+		// Group query results by route.
+		const groupedQueryResults = groupBy( sortedQueryResults, "route" );
+
+		// Sort route groups by max hits of single field, highest number of max hits first.
+		// Ie. if group A contains a field with 2 hits and group B contains 2 fields with 1 hit, group A will be sorted first.
+		const sortedGroupedQueryResults = values( groupedQueryResults ).sort( ( a, b ) => {
+			const aMaxHits = reduce( a, ( maxHitsAcc, queryResult ) => max( [ maxHitsAcc, queryResult.hits ] ), 0 );
+			const bMaxHits = reduce( b, ( maxHitsAcc, queryResult ) => max( [ maxHitsAcc, queryResult.hits ] ), 0 );
+			return bMaxHits - aMaxHits;
+		} );
+
+		setResults( sortedGroupedQueryResults );
+	}, 100 ), [ queryableSearchIndex, searchIndex ] );
+
+	const handleQueryChange = useCallback( event => {
 		setQuery( event.target.value );
-	}, 500 ), [ items ] );
+		debouncedSearch( event.target.value );
+	}, [ setQuery, debouncedSearch ] );
 
 	return <>
 		<button
 			className="yst-w-full yst-flex yst-items-center yst-bg-white yst-text-sm yst-leading-6 yst-text-slate-500 yst-rounded-md yst-border yst-border-slate-300 yst-shadow-sm yst-py-1.5 yst-pl-2 yst-pr-3 focus:yst-outline-none focus:yst-ring-2 focus:yst-ring-offset-2 focus:yst-ring-primary-500"
-			// onClick={ setOpen }
+			onClick={ setOpen }
 		>
 			<SearchIcon
 				className="yst-flex-none yst-w-5 yst-h-5 yst-mr-3 yst-text-slate-400"
 				{ ...ariaSvgProps }
 			/>
-			<span>{ __( "Quick search...", "wordpress-seo" ) }</span>
+			<span>{ query || __( "Quick search...", "wordpress-seo" ) }</span>
 		</button>
 		<Modal
 			onClose={ setClose }
```

Comment:
Important: The function should be `setOpen` to toggle the modal. `setClose` gives an error and a blank screen.

Reply:
I'm not sure what you mean? The `setClose` function sets the `isOpen` to `false`. Why would that give an error?